coordinator/internal/transitengine/transitengine.go

```diff
@@ -0,0 +1,314 @@
+func writeJSONResponse(w http.ResponseWriter, payload any) error {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	response := map[string]interface{}{
+		"data": payload,
+	}
+	return json.NewEncoder(w).Encode(response)
+}
+
+// parseEncryptionRequest returns the given HTTP request body as encryptionRequest holding
+// the b64Plaintext and the optional query parameter associatedData.
+func parseEncryptionRequest(r *http.Request) (encryptionRequest, error) {
+	defer r.Body.Close()
+	if err := validateContentType(r); err != nil {
+		return encryptionRequest{}, err
+	}
+	var encReq encryptionRequest
+	if err := json.NewDecoder(r.Body).Decode(&encReq); err != nil {
+		return encryptionRequest{}, err
+	}
+	return encReq, nil
+}
+
+// parseDecryptionRequest returns the given HTTP request body as decryptionRequest holding
+// the ciphertextContainer and the optional query parameter associatedData.
+// Checks that the URL version matches the received ciphertext request version.
+func parseDecryptionRequest(r *http.Request, version int) (decryptionRequest, error) {
+	defer r.Body.Close()
+	if err := validateContentType(r); err != nil {
+		return decryptionRequest{}, err
+	}
+	var decReq decryptionRequest
+	if err := json.NewDecoder(r.Body).Decode(&decReq); err != nil {
+		return decryptionRequest{}, err
+	}
+	if version != decReq.ciphertextContainer.version {
+		return decryptionRequest{}, errors.New("Mismatch URL and ciphertext key version")
+	}
+	return decReq, nil
+}
+
+// validateContentType ensures that if Content-Type is present, it is set to application/json.
+func validateContentType(r *http.Request) error {
+	if contentType := r.Header.Get("Content-Type"); contentType != "" {
+		if !strings.HasPrefix(contentType, "application/json") {
+			return fmt.Errorf("invalid content-type: %s (want application/json)", contentType)
+		}
+	}
+	return nil
+}
+
+// newCiphertextContainer returns a new ciphertextContainer, holding the version prefix and decoded base64 nonce and ciphertext.
+func newCiphertextContainer(encoded string) (ciphertextContainer, error) {
+	// Split "vault:vX:base64" format
+	parts := strings.SplitN(encoded, ":", 3)
+	if len(parts) < 3 {
+		return ciphertextContainer{}, fmt.Errorf("invalid ciphertext format")
+	}
+	version, err := extractVersion(parts[1])
+	if err != nil {
+		return ciphertextContainer{}, fmt.Errorf("ciphertext version: %w", err)
+	}
+	fullCiphertext, err := decodeBase64(parts[2])
+	if err != nil {
+		return ciphertextContainer{}, fmt.Errorf("decoding ciphertext: %w", err)
+	}
+
+	return ciphertextContainer{
+		version:    version,
+		nonce:      fullCiphertext[:aesGCMNonceSize],
+		ciphertext: fullCiphertext[aesGCMNonceSize:],
+	}, nil
+}
+
+// MarshalJSON wraps b64Plaintext in an object with the key "plaintext".
+func (p b64Plaintext) MarshalJSON() ([]byte, error) {
+	return json.Marshal(map[string]string{
+		"plaintext": b64.StdEncoding.EncodeToString(p),
```

Comment:
The `encryptionRequest` already maps `plaintext: b64Plaintext`. Does this now marshal as?

```json
{ "plaintext": {"plaintext": "EAWEUng1BEDFQ"}}
```
I guess it does not matter simply because we're only unmarshalling?